### PR TITLE
Aggressive `count` in BooleanWeight

### PR DESCRIPTION
**Pull request by LuXugang**
When BooleanQuery is pure disjunction, if at least one clause could match all docs, then we could get right `count`  even though there was other clause whose `count` is  unknown.

### CODE REVIEW

**Review by jpountz — COMMENTED** [review]
Thanks, I left some comments but the change makes sense to me.

File: lucene/core/src/java/org/apache/lucene/search/BooleanWeight.java [hunk 1]
```diff
@@ -470,14 +470,19 @@ private int reqCount(LeafReaderContext context) throws IOException {
...
       }
       int count = weightedClause.weight.count(context);
-      if (count == -1 || count == numDocs) {
-        // If any of the clauses has a number of matches that is unknown, the number of matches of
-        // the disjunction is unknown.
+      if (count == -1) {
+        // If one clause has a number of matches that is unknown, let's be more aggressive to check
+        // whether remain clauses could match all docs.
+        unknownCount = true;
+        continue;
```
[the remark this answers is not in the document]
Comment (LuXugang) [reply lucene/core/src/java/org/apache/lucene/search/BooleanWeight.java:483]: addressed in  https://github.com/apache/lucene/pull/12017/commits/dc13ae7bddc3ca272af7c45dd998258984469904

File: lucene/core/src/java/org/apache/lucene/search/BooleanWeight.java [hunk 2]
```diff
@@ -492,7 +497,10 @@ private int optCount(LeafReaderContext context, Occur occur) throws IOException
         return -1;
```
[the remark this answers is not in the document]
Comment (LuXugang) [reply lucene/core/src/java/org/apache/lucene/search/BooleanWeight.java]: Thanks @jpountz , you are right. addressed in  https://github.com/apache/lucene/pull/12017/commits/477b2a4e1f95f408562fc1745979bf9815ffaa75

File: lucene/core/src/test/org/apache/lucene/search/TestBooleanQuery.java [hunk 3]
```diff
@@ -1015,6 +1015,80 @@ public void testDisjunctionRandomClausesMatchesCount() throws Exception {
     }
   }
 
+  public void testAggressiveMatchCount() throws IOException {
```
[the remark this answers is not in the document]
Comment (LuXugang) [reply lucene/core/src/test/org/apache/lucene/search/TestBooleanQuery.java:1074]: addressed in https://github.com/apache/lucene/pull/12017/commits/272dec54a59dedc666cf8311c09830c94b3c5369

File: lucene/core/src/test/org/apache/lucene/search/TestBooleanQuery.java [hunk 4]
```diff
@@ -1015,6 +1015,80 @@ public void testDisjunctionRandomClausesMatchesCount() throws Exception {
...
+            .add(unknownCountQuery, Occur.MUST_NOT)
+            .add(new TermQuery(new Term("string", "abc")), Occur.MUST_NOT)
+            .build();
+    weight = searcher.createWeight(query, ScoreMode.COMPLETE, 1f);
+    // count of the first MUST_NOT clause is unknown, though the second MUST_NOT clause matche one
+    // doc, we can't figure out the number of
+    // docs
+    assertEquals(-1, weight.count(reader.leaves().get(0)));
+
+    // test pure disConjunction
```
[the remark this answers is not in the document]
Comment (LuXugang) [reply lucene/core/src/test/org/apache/lucene/search/TestBooleanQuery.java]: addressed in https://github.com/apache/lucene/pull/12017/commits/272dec54a59dedc666cf8311c09830c94b3c5369

**Review by jpountz — APPROVED** [review]

**Pull request merged**